src/TaggableCachePoolTest.php

```diff
@@ -292,6 +275,20 @@ public function testInvalidateTags()
         $this->cache->save($item);
         $this->cache->invalidateTags(['tag1']);
 
-        $this->assertTrue($this->cache->hasItem('key'), 'Item key list should be removed when clearing the tags');
+        $this->assertTrue($this->cache->hasItem('key'), 'Item k list should be removed when clearing the tags');
+    }
+
+    /**
+     * When an item is overwritten we need to clear tags for original item.
+     */
+    public function testTagsAreCleanedOnSave()
+    {
+        $pool = $this->cache;
+        $i    = $pool->getItem('key');
+        $pool->save($i->addTag('foo'));
+        $i = $pool->getItem('key');
+        $pool->save($i->addTag('bar'));
+        $pool->invalidateTags(['foo']);
+        $this->assertTrue($pool->getItem('key')->isHit());
```

Comment:
*when* it does have some sense for some particular need, the one should use first getTags, and use the result to set tags again. But this shouldn't be the default IMHO.